sys/include/configuration.h

```diff
@@ -0,0 +1,458 @@
+ *
+ * This is called by @ref configuration_verify() for the handler.
+ *
+ * @param[in]           handler     Reference to the handler
+ * @param[in]           key         Configuration key which belongs to the configuration handler
+ * @param[in]           next        Comes after the key and the handler knows how to process
+ *
+ * @return  0 on success
+ */
+typedef int (*conf_data_verify_handler) (const struct conf_handler *handler,
+                                         char *key, char *next);
+
+/**
+ * @brief   Handler prototype to apply the internal representation of a configuration item to
+ *          the configuration subject, e.g. a sensor
+ *
+ * This is called by @ref configuration_apply() for the handler.
+ * @ref conf_handler_ops_t::set only stores the configuration value to the internal location, and
+ * @ref conf_data_apply_handler::apply applies the value to the subject.
+ *
+ * @param[in]           handler     Reference to the handler
+ * @param[in]           key         Configuration key which belongs to the configuration handler
+ * @param[in]           next        Comes after the key and the handler knows how to process
+ *
+ * @return  0 on success
+ */
+typedef int (*conf_data_apply_handler) (const struct conf_handler *handler,
+                                         char *key, char *next);
+
+/**
+ * @brief   Configuration handler operations
+ */
+typedef struct conf_handler_ops {
+    conf_data_set_handler set;              /**< Set a configuration value */
+    conf_data_get_handler get;              /**< Get a configuration value */
+    conf_data_import_handler import;        /**< Import a configuration value
+                                                 from persistent storage */
+    conf_data_export_handler export;        /**< Export a configuration value
+                                                 to persistent storage */
+    conf_data_delete_handler delete;        /**< Delete a configuration value
+                                                 from persistent storage */
+    conf_data_verify_handler verify;        /**< Verify the currently set configuration */
+    conf_data_apply_handler apply;          /**< Apply the currently set configuration */
+} conf_handler_ops_t;
+
+/**
+ * @brief   An intermediate node in the configuration tree
+ */
+typedef struct conf_handler_node {
+    list_node_t node;                       /**< Every node is in a list, managed by its parent */
+    struct conf_handler_node *sub_nodes;    /**< Every node has a list of subnodes */
+    const char *subtree;                    /**< A segment in a key to a configuration item */
+    unsigned level;                         /**< Level in the configuration tree (root = 0) */
+} conf_handler_node_t;
+
+/**
+ * @brief   Static initializer for an intermediate handler node
+ *
+ * @param   path        Configuration path segment, e.g. "bar" in "foo/bar/bazz"
```

Comment:
Yes `foo` could be the parent and `bazz` could be one child.
The decision to not put the full path `foo/bar/bazz` in the handler node I made because I wanted to avoid to store the full key `foo/bar/bazz/` for example for `foo/bar/bazz/a`, `foo/bar/bazz/b`, `foo/bar/bazz/c`. I also had the idea that the tree could in fact be a list of only handler nodes which then must store the full key. This would simplify the handler lookup procedure but depending on the number of items with a common subkey string I think we would have to store many `const char *` with a common prefix substring.